src/compiler/checker.ts

```diff
@@ -5851,6 +5879,67 @@ namespace ts {
             return links.resolvedType;
         }
 
+        /**
+         * Since the source of spread types are object literals and type literals, which are not binary,
+         * this function should be called in a left folding style, with left = previous result of getSpreadType
+         * and right = the new element to be spread.
+         */
+        function getSpreadType(left: Type, right: Type, symbol: Symbol): ResolvedType {
+            Debug.assert(!!(left.flags & TypeFlags.Object) && !!(right.flags & TypeFlags.Object), "Only object types may be spread.");
+            const members = createMap<Symbol>();
+            const skippedPrivateMembers = createMap<boolean>();
+            let stringIndexInfo: IndexInfo;
+            let numberIndexInfo: IndexInfo;
+            if (left === emptyObjectType) {
+                // for the first spread element, left === emptyObjectType, so take the right's string indexer

```

Comment:
It's treated like any other object type. `emptyObjectType` is only a marker of the start of the spread type because the `checkObjectLiteral` uses `getSpreadType` in a left fold manner, and `emptyObjectType` is initial value of the fold.